File: nodejs/base/api/lib/routing/defaults/routes/default-query-builder.ts
import { McmaResource, getTableName } from "@mcma/core";
import { DocumentDatabaseTableProvider, getFilterExpressionFromKeyValuePairs, FilterExpression } from "@mcma/data";

import { DefaultRouteHandlerConfigurator } from "../default-route-handler-configurator";
import { DefaultRouteBuilder } from "../default-route-builder";

export function defaultQueryBuilder<T extends McmaResource>(
    dbTableProvider: DocumentDatabaseTableProvider,
    root: string
): DefaultRouteBuilder<T[]> {
    return new DefaultRouteBuilder<T[]>(
        "GET",
        root,
        new DefaultRouteHandlerConfigurator<T[]>((onStarted?, onCompleted?) => 
            async (requestContext) => {
                if (onStarted) {
                    const continueRequest = await onStarted(requestContext);
                    if (continueRequest !== undefined && !continueRequest) {
                        return;
                    }
                }

                const filterExpression: FilterExpression<T> =
                    requestContext.request.queryStringParameters && Object.keys(requestContext.request.queryStringParameters).length > 0
                        ? getFilterExpressionFromKeyValuePairs<T>(requestContext.request.queryStringParameters)
                        : null;

                let pageNumber = 0;
                if (requestContext.request.queryStringParameters["pageNumber"]) {
                    pageNumber = parseInt(requestContext.request.queryStringParameters["pageNumber"]);
                    if (isNaN(pageNumber)) {
                        pageNumber = 0;
                    }
                }

                let pageSize: number;
                if (requestContext.request.queryStringParameters["pageSize"]) {
                    pageSize = parseInt(requestContext.request.queryStringParameters["pageSize"]);
                    if (isNaN(pageSize)) {
                        pageSize = null;
                    }
                }

                const table = await dbTableProvider.get(getTableName(requestContext));

                const resources = await table.query<T>({ path: requestContext.request.path, pageNumber, pageSize, filterExpression });
                if (onCompleted) {
                    await onCompleted(requestContext, resources);
                }

                requestContext.setResponseBody(resources);
            }
        )
    );
}

File: dotnet/base/Mcma.Api/Routing/Defaults/Routes/DefaultQueryBuilder.cs

using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using Mcma.Context;
using Mcma.Data;
using Mcma.Data.DocumentDatabase.Queries;

namespace Mcma.Api.Routing.Defaults.Routes
{
    public static class DefaultQueryBuilder
    {
        public static DefaultRouteBuilder<IEnumerable<TResource>> Get<TResource>(
            IDocumentDatabaseTableProvider dbTableProvider,
            string root)
            where TResource : McmaResource
            => 
            new DefaultRouteBuilder<IEnumerable<TResource>>(
                HttpMethod.Get,
                root, 
                new DefaultRouteHandlerConfigurator<IEnumerable<TResource>>(
                    (onStarted, onCompleted) =>
                        async requestContext =>
                        {
                            // invoke the start handler, if any
                            if (onStarted != null)
                                if (!await onStarted(requestContext))
                                    return;

                            var table = await dbTableProvider.GetAsync(requestContext.TableName());

                            // get all resources from the table, applying in-memory filtering using the query string (if any)
                            var resources = (await table.QueryAsync(requestContext.ToQuery<TResource>())).ToArray();

                            // invoke the completion handler with the results
                            if (onCompleted != null)
                                await onCompleted(requestContext, resources);

                            // return the results as JSON in the body of the response
                            // NOTE: This will never return a 404 - just an empty collection
                            requestContext.SetResponseBody(resources);
                        }));
    }
}